modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/distributed/near/IgniteCacheDistributedQueryDefaultTimeoutSelfTest.java

```diff
@@ -0,0 +1,199 @@
+
+        CacheConfiguration<Integer, String> ccfg = new CacheConfiguration<>(DEFAULT_CACHE_NAME);
+        ccfg.setIndexedTypes(Integer.class, String.class);
+
+        cfg.setCacheConfiguration(ccfg);
+
+        if ("client".equals(igniteInstanceName))
+            cfg.setClientMode(true);
+
+        cfg.setDefaultQueryTimeout(DEFAULT_QUERY_TIMEOUT);
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        G.ignite().cache(DEFAULT_CACHE_NAME).removeAll();
+    }
+
+    /** */
+    @Test
+    public void testRemoteQueryExecutionTimeout() throws Exception {
+        testQueryCancel(CACHE_SIZE, VAL_SIZE, QRY_1, 500, TimeUnit.MILLISECONDS, true, true);
+    }
+
+    /** */
+    @Test
+    public void testRemoteQueryWithMergeTableTimeout() throws Exception {
+        testQueryCancel(CACHE_SIZE, VAL_SIZE, QRY_2, 500, TimeUnit.MILLISECONDS, true, false);
+    }
+
+    /** */
+    @Test
+    public void testRemoteQueryExecutionCancel0() throws Exception {
+        testQueryCancel(CACHE_SIZE, VAL_SIZE, QRY_1, 1, TimeUnit.MILLISECONDS, false, true);
+    }
+
+    /** */
+    private void testQueryCancel(int keyCnt, int valSize, String sql, int timeoutUnits, TimeUnit timeUnit,
+        boolean timeout, boolean checkCanceled) throws Exception {
+        try (Ignite client = startGrid("client")) {
+            IgniteCache<Object, Object> cache = client.cache(DEFAULT_CACHE_NAME);
+
+            assertEquals(0, cache.localSize());
+
+            int p = 1;
+            for (int i = 1; i <= keyCnt; i++) {
+                char[] tmp = new char[valSize];
+                Arrays.fill(tmp, ' ');
+                cache.put(i, new String(tmp));
+
+                if (i/(float)keyCnt >= p/10f) {
+                    log().info("Loaded " + i + " of " + keyCnt);
+
+                    p++;
+                }
+            }
+
+            assertEquals(0, cache.localSize());
+
+            SqlFieldsQuery qry = new SqlFieldsQuery(sql);
+
+            final QueryCursor<List<?>> cursor;
+            if (timeout) {
+                qry.setTimeout(timeoutUnits, timeUnit);
+
+                cursor = cache.query(qry);
+            }
+            else {
+                cursor = cache.query(qry);
+
+                client.scheduler().runLocal(new Runnable() {
+                    @Override public void run() {
+                        cursor.close();
+                    }
+                }, timeoutUnits, timeUnit);
+            }
+
+            try (QueryCursor<List<?>> ignored = cursor) {
+                cursor.getAll();
+
+                if (checkCanceled)
+                    fail("Query not canceled");
+            }
+            catch (CacheException ex) {
+                error("Got expected exception", ex);
+
+                assertNotNull("Must throw correct exception", X.cause(ex, QueryCancelledException.class));
```

Comment:
Similarly, I do not  see how do we distinguish _manual cancel_ vs _explicit query timeout_ vs _default query timeout here_.

Reply:
If you note the tests in IgniteCacheDistributedQueryStopOnCancelOrTimeoutSelfTest also, it is similar to Default Query timeout tests. I suppose the reason being we are throwing QueryCancelledException for either manual cancel or query being cancelled due to timeout. I agree throwing a separate exception for timeout vs manual cancel will be better approach to handle different scenarios and help with logging and exception handling.

The only way to differentiate the explicit query timeout vs default query timeout is like in `IgniteCacheLocalQueryDefaultTimeoutSelfTest.testQueryDefaultTimeout() test the method testQuery() method is called with param timeout as false and it will result in query cancellation due to default query timeout.`